colorize_unix.go

```diff
@@ -3,25 +3,20 @@
 package hclog
 
 import (
+	"os"
+
 	"github.com/mattn/go-isatty"
 )
 
-// setColorization will mutate the values of this logger
-// to approperately configure colorization options. It provides
-// a wrapper to the output stream on Windows systems.
-func (l *intLogger) setColorization(opts *LoggerOptions) {
-	switch opts.Color {
-	case ColorOff:
-		fallthrough
-	case ForceColor:
+func withColor(w *writer) {
+	switch w.color {
+	case ColorOff, ForceColor:
 		return
 	case AutoColor:
-		fi := l.checkWriterIsFile()
-		isUnixTerm := isatty.IsTerminal(fi.Fd())
-		isCygwinTerm := isatty.IsCygwinTerminal(fi.Fd())
-		isTerm := isUnixTerm || isCygwinTerm
-		if !isTerm {
-			l.writer.color = ColorOff
+		fd := os.Stdout.Fd()
```

Comment:
I don't see why this would be a valid change, why should what `os.Stdout` can do influence `w.w` without checking? The user might be sending the data to a `bytes.Buffer` and be depending on the fact that AutoColor would be disabled because it's not something that supports color.

Reply:
That scenario (of passing in a buffer as the `io.Writer`) is exactly the problem this PR is attempting to fix, but I guess it did not explain it well.

This problem can be seen in Consul. This line (https://github.com/hashicorp/consul/blob/v1.9.1/command/agent/agent.go#L163) builds an `io.Writer` that is [used as the writer for logging](https://github.com/hashicorp/consul/blob/v1.9.1/agent/setup.go#L66).

The current implementation of `setColorization` in hclog prevents Consul from using colors because it expects the `io.Writer` to be an `os.File` and a tty. In the case of Consul it is neither, but it does *wrap* `os.Stdout`, so color codes would work if they were enabled. It is possible that I could make substantial changes to Consul to unwrap `os.Stdout`, but I don't think that is the correct fix.

I think the problem is that `hclog` makes some incorrect assumptions. The `io.Writer` it receives is not really relevant. An application could buffer the logs, and then write them to stdout and expect to see color. With the changes I made in this PR that scenario is now possible (where as previously it was impossible to get color output). 

As you've pointed out this does introduce a different problem. If someone is using hclog with an `os.Stdout` that is a valid tty, and they do not want color, they must explicitly disable color. I think this is the correct behaviour. Auto detect will generally be correct, and in any case where the caller knows they explicitly do not want color they have the option to turn it off by setting `LoggerOptions.Color = ColorOff`.

I think the important part of this change is the removal of `checkWriterIsFile`. Another way to solve this would to to keep checking the `w.w` writer, but only in the cause when color is set to  `ColorAuto`, and never panic when `ColorForce` is set. That would allow Consul to detect the tty itself, and set `ColorForce`.

Do you prefer that option?

Reply:
In your case, why not set `ForceColor` with consul? I'd rather not change the existing behavior to have color suddenly be on when someone passes in a buffer.

I disagree that `if os.Stdout supports color, then everything is colorized by default` is the right behavior. 

`checkWriterIsFile` could instead see if `w.w` supports `Fd() uintptr` and if so, check if the returned fd is a terminal. That would be more go-ish.